nemo/export/quantize/quantizer.py

```diff
@@ -209,38 +135,60 @@ def _sample_output(model):
             },
         )
 
-        if torch.distributed.get_rank() == 0:
-            print(f'Example NeMo output after PTQ: {response["sentences"]}"')
+        logging.info(f'Example NeMo output before export: {response["sentences"]}"')
 
+    @staticmethod
     def quantize(
-        self,
-        model_file: str,
-        dataloader: Optional[List[List[str]]],
-        tensor_model_parallel_size: Optional[int] = None,
-        pipeline_model_parallel_size: Optional[int] = None,
+        model: MegatronGPTModel,
+        forward_loop: Callable[[MegatronGPTModel], None],
+        quantization_config: DictConfig,
+        inference_config: Optional[DictConfig] = None,
     ):
-        """Quantize model checkpoint using given dataloader and optional custom parallelism settings."""
-        model = self._load_model(model_file, tensor_model_parallel_size, pipeline_model_parallel_size)
+        """Quantize model checkpoint using given dataloader.
 
-        if self.quantization_config.algorithm is None:
-            return model
-
-        model.set_inference_config(OmegaConf.to_container(self.inference_config))
+        Expected keys in `quantization_config`:
+            - algorithm: str
+            - decoder_type: str
+            - awq_block_size: int (only for awq algorithms)
+        """
+        if not HAVE_MODELOPT:
+            raise RuntimeError("nvidia-modelopt is needed to use Quantizer") from HAVE_MODELOPT_ERROR
 
-        def forward_loop(model):
-            print("Calibrating the model...")
-            for i, batch in enumerate(tqdm(dataloader)):
-                model.predict_step(batch, i)
+        Quantizer._setup(model, inference_config)
+
+        algorithm = quantization_config.algorithm
+        assert algorithm in QUANT_CFG_CHOICES
+        quant_cfg = QUANT_CFG_CHOICES[algorithm]
+        decoder_type = quantization_config.decoder_type
+
+        logging.info(f"Quantizing model to {algorithm}...")
+
+        if "awq" in algorithm:
+            weight_quantizer = quant_cfg["quant_cfg"]["*weight_quantizer"]
+            if isinstance(weight_quantizer, list):
+                weight_quantizer = weight_quantizer[0]
+            weight_quantizer["block_sizes"][-1] = quantization_config.awq_block_size
+
+        # Always turn on FP8 kv cache to save memory footprint.
+        # For int8_sq, we use int8 kv cache.
+        # TODO: Investigate why enabling FP8 kv cache will cause accuracy regressions for Nemotron.
+        enable_quant_kv_cache = "int8" not in algorithm and decoder_type != "gptnext"
```

Comment:
@janekl Should this instead be like below:

```suggestion
        enable_quant_kv_cache = "fp8" in algorithm and decoder_type != "gptnext"
```

Since the comment says to enable only for fp8 but the code enable for non-int8 hence int4 also does the kv cache quantization which makes the results worse

Reply:
1. Whatever we do it should be exactly the same as in [nemo_ptq.py](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/06a155388edd4a240051176a67a11886b15db082/llm_ptq/nemo_ptq.py#L237-L239) from modelopt
2. Could you @realAsma and/or @cjluo-omniml comment on this proposal?

Reply:
"comment says to enable only for fp8" -> That is not correct. Comments says to enable fp8 for all formats except int8_sq. So yes, int4 quantization will use fp8 kv cache quantization. For most of the models fp8 kv cache does not cause much accuracy degradation.

Reply:
INT4 has a little accuracy degradation if we enable KV Cache hence I though we didn't want to enable it. PTQ val lm loss with KV cache enabled is 1.35 vs 1.31 and QAT was around 1.2 instead of 1.17 (don't exactly remember QAT numbers)

Reply:
Thanks for catching this. The change is fine but requires MMLU accuracy check for, say, Llama3 (8b, 70b) int4_awq.

It's quite easy for me to test so I'll handle that

Reply:
...but first let me double check: @realAsma I conclude from your comment that you suggest keeping the current behavior, right?

Only that comment "comment says to enable only for fp8" needs to be changed then.

Reply:
I ran tests for enabling / disabling KV cache and it looks it doesn't really matter much for accuracy.
```
     MODEL  QFORMAT  GPUS DTYPE KV_CACHE  MMLU  LAMBADA  TOKENS_PER_SEC
 LLAMA3-8B int4_awq     1  fp16   Enable 0.628   0.662          2045.11
 LLAMA3-8B int4_awq     1  fp16  Disable 0.628   0.663          2037.57
LLAMA3-70B int4_awq     1  fp16   Enable 0.779   0.727           459.53
LLAMA3-70B int4_awq     1  fp16  Disable 0.778   0.729              NaN
```
benchmark.py OOM-ed for LLAMA3-70B with KV_CACHE = Disable

Reply:
Can you try non-AWQ `int4` as well please?

Reply:
Let's postpone this so as to not to block this MR. You could optionally open some issue / ticket to track it.

Please also modify the comment "# Always turn on FP8 kv cache to save memory footprint." to match the implementation as we discussed

Reply:
As Asma mentioned, the comment is correct. KV cache is FP8 quantized regardless of the quantization algorithm. 
Whether we enable fp8 kv cache for int4 model - that I will bring up in the qat PR